src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+                       contract_name TEXT,
+                       map_name TEXT,
+                       key_type TEXT,
+                       value_type TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS data_table
+                      (data_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       map_identifier INTEGER,
+                       key TEXT,
+                       value TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS contracts
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       contract_data TEXT)",
+                            NO_PARAMS)?;
+
+        contract_db.check_schema()?;
+
+        Ok(contract_db)
+    }
+
+    pub fn open(filename: &str) -> Result<SqliteContractDatabase> {
+        let contract_db = SqliteContractDatabase::inner_open(filename)?;
+
+        contract_db.check_schema()?;
+        Ok(contract_db)
+    }
+
+    fn check_schema(&self) -> Result<()> {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        if let Some(ref conn) = self.conn {
+            let _: String = conn.query_row(sql, &["maps_table"],
+                                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["contracts"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["data_table"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            Ok(())
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+
+    fn inner_open(filename: &str) -> Result<SqliteContractDatabase> {
+        let conn = Connection::open(filename)
+            .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+        Ok(SqliteContractDatabase {
+            conn: Some(conn),
+            save_point: 0
+        })
+    }
+
+    fn load_map(&self, contract_name: &str, map_name: &str) -> Result<SqliteDataMap> {
+        let (map_identifier, key_type, value_type): (_, String, String) =
+            self.query_row(
+                "SELECT map_identifier, key_type, value_type FROM maps_table WHERE contract_name = ? AND map_name = ?",
+                &[contract_name, map_name],
+                |row| {
+                    (row.get(0), row.get(1), row.get(2))
+                })?
+            .ok_or(Error::new(ErrType::UndefinedMap(map_name.to_string())))?;
+
+        Ok(SqliteDataMap {
+            map_identifier: map_identifier,
+            key_type: TypeSignature::deserialize(&key_type)?,
```

Comment:
I'm wondering if using the `?` operator is the right thing here -- if the database somehow gets unparseable data inserted into it, should the node `panic!`?

Reply:
Right -- see the previous comment about the BadPrincipal error. Something should happen here that's `panic!` like, but I wasn't sure if we wanted to have some higher level handler that did it's own logging or cleanup, etc. I think we'll need to differentiate between, say, runtime errors in the contract code (which will result in the tx fee still be assessed, the chain advancing) and implementation or deployment errors on the node (which should result in the node dying).

Reply:
What if we switched these cases to `panic!` for now, and later on, added a custom `panic!`-like macro?  I'm just trying to be diligent about ensuring that fatal errors like this actually do kill the node, since the only thing worse than these kinds of errors happening is the node _not_ dying.

Reply:
Okay, that works for me.